### PR TITLE
[42873] Manage "OAuth applications" for storages from within storages administration

### PR DESCRIPTION
**Pull request by wielinde**
https://community.openproject.org/projects/openproject/work_packages/42873

Done:

- Extend data model with a polymorphic relation between `::Doorkeeper::Application` (aka "OAuth application") and storages called "integration".
  - Migration
  - Patch `::Doorkeeper::Application` with association 
- Extend Storages `CreateService` to also create `::Doorkeeper::Application`
- Extend Storages `UpdateService` to update the `::Doorkeeper::Application` (`name` and `redirect_uri`)
- Extend OAuth application `ApplicationContract` to validate completeness of the "integration" relation

Todo:

- [ ] ~~Reveal correct secret and do not shorten (as it would shorten the hash... non sense)~~
  - [x] Hide secrets (discussed with product team)
- [x] Fixing "copy to clipboard" when showing OpenProject OAuth details.
- [x] Add Info/Warning box when showing OpenProject OAuth details. Check Figma.
- [ ] ~~Generaly, check wording with Figma~~ (moved: https://community.openproject.org/projects/openproject/work_packages/43111/activity)
- [x] Extend feature specs to cover creation of OAuth application

### CODE REVIEW

File: frontend/src/global_styles/content/_toast.sass [hunk 1]
```diff
@@ -173,6 +173,8 @@ $nm-upload-box-padding: rem-calc(15) rem-calc(25)
   & p, & div, & ul>li, & .button
     font-size: $nm-font-size
 
+  &.-with-bottom-spacing
+    margin-bottom: 2rem
```
Comment (apfohl) [inline frontend/src/global_styles/content/_toast.sass:177]: @HDinger I would like your opinion on this. Can we do it like this? We needed to make some room there.

Comment (HDinger) [reply frontend/src/global_styles/content/_toast.sass:177]: What is the case in which that is needed? It seems a bit odd to me as the toast should be floating above all elements and thus would not need additional spacing to the bottom. 
If it is inevitable to add such a rule, I'd change the class name. With out new DS based approach we try to shift to naming modifier classes contextual (e.g. `-error`, `-blocked`, etc.). Thus every developer knows directly when to use this class.

Comment (apfohl) [reply frontend/src/global_styles/content/_toast.sass:177]: I agree. We need to use it here. So it is not a toast in general, but the same UI component.

image

Comment (HDinger) [reply frontend/src/global_styles/content/_toast.sass:177]: And why isn't that a normal toast that floats at the top of the page?

Comment (apfohl) [reply frontend/src/global_styles/content/_toast.sass:177]: Sorry, I haven't questioned this. Maybe @marcalcobe can step in here, as I don't know.

Comment (wielinde) [reply frontend/src/global_styles/content/_toast.sass:177]: Let me add to that discussion: If you search our code base for `toast -info` you find a couple of places where we are using these boxes (that were renamed to toast), as wrappers for highlighted content. That was a very normal use case and I argue it still is. These are use cases that are unlike flash messages (i.e. validation errors, creation success, ...).

File: modules/storages/app/views/storages/admin/storages/show.html.erb [hunk 2]
```diff
@@ -134,7 +134,7 @@ See COPYRIGHT and LICENSE files for more details.
       <div class="attributes-key-value--key"><%= t("storages.provider_types.#{@object.provider_type}") %> <%= t(:'storages.label_oauth_client_secret') %></div>
       <div class="attributes-key-value--value-container">
         <div class="attributes-key-value--value -text">
-          <span><%= short_secret(@object.oauth_client.client_secret) %></span>
+          <span>●●●●●●●●●●●●●●●●</span>
```
Comment (wielinde) [inline modules/storages/app/views/storages/admin/storages/show.html.erb]: @apfohl Is it necessary to remove the shortening of the Nextcloud OAuth client secret? Here, the value is not hashed in our DB so we technically can reveal some characters. I argue that it increases the UX for the user as she can still compare the first two and the last two letters.

File: modules/storages/spec/features/admin_storages_spec.rb [hunk 3]
```diff
@@ -156,9 +162,16 @@
...
+    # Check created OAuth2 Application
+    page.find('[data-qa-selector="op-menu--item-action"]', text: 'Authentication').click
+    page.find('[data-qa-selector="op-menu--item-action"]', text: 'OAuth applications').click
+    expect(page).to have_text("Other NC (#{storage_type})")
+    expect(page).not_to have_selector('a', text: 'Delete')
+
+    # Editing oauth app opens storage edit controller
+    page.find('a', text: 'Edit').click
+    expect(page).to have_title("Edit: Other NC")
+    page.find('button[type=submit]', text: "Save").click
```
Comment (wielinde) [inline modules/storages/spec/features/admin_storages_spec.rb]: @apfohl Apart from the "Edit" button, covering the link on the name is missing here, I believe. It should lead to the show view of that storage.

File: modules/storages/app/services/storages/oauth_applications/create_service.rb [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+#++
+
+# The logic for creating storage was extracted from the controller and put into
+# a service: https://dev.to/joker666/ruby-on-rails-pattern-service-objects-b19
+# Purpose: create and persist a Storages::Storage record
+# Used by: Storages::Admin::StoragesController#create, could also be used by the
+# API in the future.
+# Reference: https://www.openproject.org/docs/development/concepts/contracted-services/
+# The comments here are also valid for the other *_service.rb files
+module Storages::OAuthApplications
```
Comment (wielinde) [inline modules/storages/app/services/storages/oauth_applications/create_service.rb:36]: This file is not yet covered with a spec. The spec is missing.

**Review by wielinde — COMMENTED** [review]
@apfohl I first started commenting inline and only later started to create a whole review. So please also consider the inline comments, too.

File: modules/storages/app/views/storages/admin/storages/show.html.erb (hunk 2 above)
Comment (apfohl) [reply modules/storages/app/views/storages/admin/storages/show.html.erb]: I haven't decided this and need you to talk to @marcalcobe. I will change it to what it should be afterwards.

**Comment by apfohl** [comment]
I always do. They are part of your review. This comment works as some sort of summary for your review.

File: modules/storages/app/services/storages/oauth_applications/create_service.rb (hunk 4 above)
Comment (wielinde) [reply modules/storages/app/services/storages/oauth_applications/create_service.rb:36]: I just pushed a spec covering that service.

File: frontend/src/global_styles/content/_toast.sass (hunk 1 above)
Comment (apfohl) [reply frontend/src/global_styles/content/_toast.sass:177]: So, I assume it is okay to add the spacing that way?

File: modules/storages/app/views/storages/admin/storages/edit.html.erb [hunk 5]
```diff
@@ -28,11 +56,11 @@
         <div class="attributes-key-value--key"><%= t("storages.provider_types.#{@object.provider_type}") %> <%= t(:'storages.label_oauth_client_secret') %></div>
         <div class="attributes-key-value--value-container">
           <div class="attributes-key-value--value -text">
-            <span><%= @short_secret %></span>
+            <span><%= short_secret(@object.oauth_client.client_secret) %></span>
```
Comment (Kharonus) [inline modules/storages/app/views/storages/admin/storages/edit.html.erb:59]: this is not what UX decided to do here. It looks weird of shortening one secret and hiding the other. We should be consistent and consistent with OP UI is hiding secrets. Why should we reveal any of the characters?

File: modules/storages/spec/features/admin_storages_spec.rb (hunk 3 above)
Comment (Kharonus) [reply modules/storages/spec/features/admin_storages_spec.rb]: I just realized that behaviour and I find it wrong. Clicking on the oauth app link should open the oauth app, not the storage view. And the delete button must be hidden there, and the edit button redirected. but it is inconsistent, that the storage is opened here.

**Pull request merged**